### PR TITLE
feat(lsp): respect nested deno.json for fmt and lint config

### PR DESCRIPTION
**Pull request by nayeemrmn**
More extraction from #21029.

### CODE REVIEW

File: cli/lsp/config.rs [hunk 1]
```diff
@@ -1427,181 +1423,193 @@ impl ConfigData {
...
       .lock()
       .as_ref()
-      .and_then(|(_, d)| d.lockfile.clone())
+      .and_then(|s| self.scopes.lock().get(s).and_then(|d| d.lockfile.clone()))
+  }
+
+  pub fn root_import_map(&self) -> Option<Arc<ImportMap>> {
+    self.first_folder.lock().as_ref().and_then(|s| {
+      self.scopes.lock().get(s).and_then(|d| d.import_map.clone())
+    })
```
Comment (dsherret) [inline cli/lsp/config.rs:1465]: We're acquiring locks in different orders, ex. `first_folder.lock()` here then `self.scopes.lock()` and then below on line 1472 it acquires `self.scopes.lock()` and then `self.first_folder.lock()`. Not sure if it's related, but that might be the cause of the hanging.

This use of multiple locks is very sketchy and we should remove it. Looking at the code a bit more, it looks like we could remove these locks entirely and then create a `ConfigTreeSnapshot` to send to tsc. Sending mutable data in the tsc snapshot is not ideal.

Comment (nayeemrmn) [reply cli/lsp/config.rs:1465]: Turns out we don't need locks at all since we can just update with a mut reference...

File: cli/lsp/config.rs [hunk 2]
```diff
@@ -1617,10 +1625,10 @@ impl ConfigTree {
       return true;
     }
     self
-      .root
+      .scopes
       .lock()
-      .as_ref()
-      .is_some_and(|(_, d)| d.watched_files.contains_key(specifier))
+      .values()
+      .any(|data| data.watched_files.contains_key(specifier))
```
Comment (nathanwhit) [inline cli/lsp/config.rs:1631]: Not sure how many scopes we expect to have, but we may want to optimize this down the line

File: tests/integration/lsp_tests.rs [hunk 3]
```diff
@@ -1061,6 +1061,7 @@ fn lsp_did_change_deno_configuration_notification() {
       }],
     }))
   );
+  client.shutdown();
```
[the remark this answers is not in the document]
Comment (nayeemrmn) [reply tests/integration/lsp_tests.rs:1064]: I think in particular it should be implemented on a guard returned by `client.initialize()`, at the same time that would be a bit more error prone. Not sure.

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**